### PR TITLE
enable fusedsdpa for prompt attention with env VLLM_PROMPT_USE_FUSEDSDPA=true

### PR DESCRIPTION
**Pull request by libinta**
The current baseline with default prompt attention, performance is slow due to softmax kernel.
Like bs1 + input 20k, device time ~8s
With enabling of fusedsdpa, we can utilize optimized kernel scaled_masked_triangular_softmax_fwd_bf16, it speeds up the performance, bs1  + input 20k, device time ~2s.

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/layers/logits_processor.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ def _get_logits(self, hidden_states: torch.Tensor, embedding: torch.Tensor,
         if embedding_bias is not None:
             logits += embedding_bias
         # NOTE(kzawora): HPU PT bridge is missing support for single-rank gather. We'll use all-gather on Gaudi for now.
-        gather_op = tensor_model_parallel_all_gather if is_hpu() else tensor_model_parallel_gather
+        gather_op = tensor_model_parallel_gather
```
Comment (kzawora-intel) [inline vllm/model_executor/layers/logits_processor.py:70]: why do we need this?

**Review by kzawora-intel — CHANGES_REQUESTED** [review]
Thanks for the PR! I've suggested some minor changes. Couple of questions:
1) Can we have this change targetting habana_main?
2) Formatting is off in some places. Can you pass it through format.sh?

File: vllm/attention/backends/habana_attn.py [hunk 2]
```diff
@@ -109,7 +109,7 @@ def __init__(
 
         assert self.num_heads % self.num_kv_heads == 0
         self.num_queries_per_kv = self.num_heads // self.num_kv_heads
-
+        self.prefill_usefusedsdpa = os.getenv('VLLM_PREFILL_USE_FUSESDAPA', '0') == '1'
```
[the remark this answers is not in the document]
Comment (libinta) [reply vllm/attention/backends/habana_attn.py:112]: will do

File: vllm/worker/habana_model_runner.py [hunk 3]
```diff
@@ -227,6 +227,7 @@ class HpuModelAdapter():
     def __init__(self, model, block_size):
         self.model = model
         self.block_size = block_size
+        self.prefill_usefusedsdpa = os.getenv('VLLM_PREFILL_USE_FUSESDAPA', '0') == '1'
```
[the remark this answers is not in the document]
Comment (libinta) [reply vllm/worker/habana_model_runner.py:230]: will do

File: vllm/model_executor/layers/logits_processor.py (hunk 1 above)
Comment (libinta) [reply vllm/model_executor/layers/logits_processor.py:70]: we don't need this for this change, it's for 8x, I will remove it

**Review by wpyszka — APPROVED** [review]
looks ok to me

File: vllm/hpu/xops.py [hunk 4]
```diff
@@ -21,23 +57,36 @@ def prompt_attention(
...
+            attn_weights = attn_weights.flatten(1, 2)
+    else:
+        #TODO: remove the handle for query_heads != kv_head for fusedsdpa after SW-195415 fix
+        if query_heads != kv_heads:
+            key = repeat_kv(key, int(query_heads//kv_heads))
+            value = repeat_kv(value, int(query_heads//kv_heads))
+        softmax_mode = 'fast'
+        recompute_mode = True
+        #attn_weights = gaudi_flash_attn_v1(query, key, value, attn_bias, softmax_mode, scale, 8192)
+        attn_weights = FusedSDPA.apply(query, key, value, None, 0.0, True, scale, softmax_mode, recompute_mode, valid_sequence_lengths, 'left')
```
Comment (libinta) [inline vllm/hpu/xops.py:90]: need to change 'left' to 'right' as vllm is right padding
